File: src/main/java/me/wickso/spark/example/WordCount.java
package me.wickso.spark.example;

import java.util.Arrays;

import org.apache.commons.lang3.ArrayUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import scala.Tuple2;

public class WordCount {
	public static void main(String[] args) {
		if (ArrayUtils.getLength(args) != 3) {
			System.err.println("Usage: WordCount <Master> <Input> <Output>");
			return;
		}

		try (JavaSparkContext sc = getSparkContext("WordCount", args[0]);) {
			JavaRDD<String> inputRDD = getInputRDD(sc, args[1]);

			JavaPairRDD<String, Integer> resultRDD = process(inputRDD);

			handleResult(resultRDD, args[2]);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	private static JavaSparkContext getSparkContext(String appName, String master) {
		SparkConf conf = new SparkConf().setAppName(appName).setMaster(master);
		return new JavaSparkContext(conf);
	}

	private static JavaRDD<String> getInputRDD(JavaSparkContext sc, String input) {
		return sc.textFile(input);
	}

	static JavaPairRDD<String, Integer> process(JavaRDD<String> inputRDD) {
		return inputRDD.flatMap(s -> Arrays.asList(s.split(" ")).iterator())
			.mapToPair(w -> new Tuple2<>(w, 1))
			.reduceByKey(Integer::sum);
	}

	private static void handleResult(JavaPairRDD<String, Integer> resultRDD, String output) {
		resultRDD.saveAsTextFile(output);
	}
}

File: src/main/scala/me/wickso/spark/example/WordCount.scala
package me.wickso.spark.example

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount {

  def main(args: Array[String]): Unit = {
    require(args.length == 3, "Usage: WordCount <Master> <Input> <Output>")

    val sc = getSparkContext("WordCount", args(0))
    val inputRDD = getInputRDD(sc, args(1))
    val resultRDD = process(inputRDD)

    handleResult(resultRDD, args(2))
  }

  private def getSparkContext(appName: String, master: String) = {
    val conf = new SparkConf().setAppName(appName).setMaster(master)
    new SparkContext(conf)
  }

  private def getInputRDD(sc: SparkContext, input: String) = sc.textFile(input)

  def process(inputRDD: RDD[String]) = inputRDD
    .flatMap(str => str.split(" "))
    .map((_, 1))
    .reduceByKey(_ + _)

  private def handleResult(resultRDD: RDD[(String, Int)], output: String) {
    resultRDD.saveAsTextFile(output);
  }
}
